deps/9999-stream-res-err.md

```diff
@@ -0,0 +1,188 @@
+
+**Required Reviewers**: [@ryanolson](https://github.com/ryanolson) [@grahamking](https://github.com/grahamking)
+
+**Review Date**: [Date for review]
+
+**Pull Request**: [Link to Pull Request of the Proposal itself]
+
+**Implementation PR / Tracking Issue**: [Link to Pull Request or Tracking Issue for Implementation]
+
+# Summary
+
+Network level errors may occur while streaming responses from the server back to the client. If they
+occur, these errors must be made available to the client response stream listener.
+
+# Motivation
+
+The client response stream listener currently does not know why a stream is closed. Normally, a
+stream is closed because the server has finished producing all the responses, but it can also be due
+to failure of the server or the network connecting the client to the server.
+
+Knowing why the stream is closed at the network/router layer is vital to the ability of detecting
+faults while the server is streaming responses back to the client.
+
+For instance, in the current
+[router implementation](https://github.com/ai-dynamo/dynamo/blob/fcfc21f20e53908cedc41a91bbd594283ecf45db/lib/runtime/src/pipeline/network/egress/addressed_router.rs#L165-L174),
+if it is unable to restore the bytes back to the original object, it cannot relay the error back to
+the client response stream consumer for proper handling, and instead it silently skips the response
+that failed to restore.
+
+## Goals
+
+N/A
+
+### Non Goals
+
+N/A
+
+## Requirements
+
+N/A
+
+# Proposal
+
+Each response streamed will be wrapped in a `Result<U, E>`, where `U` is the response type and `E`
+is the error type, for propagating errors that occur while streaming responses from the server back
+to the client response stream listener.
+
+For instance, change the
+[`Ingress<SingleIn<T>, ManyOut<U>>`](https://github.com/ai-dynamo/dynamo/blob/fcfc21f20e53908cedc41a91bbd594283ecf45db/lib/runtime/src/pipeline/network/ingress/push_handler.rs#L20)
+type to `Ingress<SingleIn<T>, ManyOut<Result<U, E>>>` and 
+[`Egress<SingleIn<T>, ManyOut<U>>`](https://github.com/ai-dynamo/dynamo/blob/fcfc21f20e53908cedc41a91bbd594283ecf45db/lib/runtime/src/pipeline/network.rs#L246-L247)
+type to `Egress<SingleIn<T>, ManyOut<Result<U, E>>>`.
+
+The `Result<U, E>` wrapper around the opaque response type `U` provides the ability for the
+network/router to relay error information back to the client response stream consumer. Without the
+additional wrapper, it is impossible for the network/router to yield an error response from a
+stream, because the response object is opaque to the network/router.
+
+The error type `E` is
+```rust
+#[derive(Serialize, Deserialize, Debug)]
+pub struct StreamError {
+    pub flags: u32,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub message: Option<String>,
+}
+```
+where each bit on the `flags` carries a boolean message. Currently, only the most significant bit is
+used:
```

Comment:
If flags is so you can have different kind of errors, is an enum with `thiserror` what you're looking for? That's what we usually use.